File: backend/src/neuromarketing/schemas/analysis.py
from datetime import datetime

from pydantic import BaseModel, Field

from neuromarketing.schemas.enums import AnalysisStatus, ProcessingStage


# --- Request/Response Models ---


class AnalysisCreateResponse(BaseModel):
    analysis_id: str
    status: AnalysisStatus
    created_at: datetime


class AnalysisStatusResponse(BaseModel):
    analysis_id: str
    status: AnalysisStatus
    stage: ProcessingStage | None = None
    stage_label: str | None = None
    progress_percent: int = 0


class VideoMetadataSchema(BaseModel):
    filename: str
    duration_seconds: float
    resolution: str
    file_size_bytes: int


# --- Report Models ---


class TimelineSegment(BaseModel):
    start_seconds: float
    end_seconds: float
    label: str
    dominant_regions: list[str]
    cognitive_state: str
    engagement_level: str = Field(pattern="^(high|medium|low)$")
    insight: str


class RegionInsight(BaseModel):
    roi_name: str
    full_name: str
    cognitive_function: str
    activation_rank: int
    marketing_implication: str


class Recommendation(BaseModel):
    category: str
    title: str
    description: str
    priority: str = Field(pattern="^(high|medium|low)$")


class AnalysisReport(BaseModel):
    executive_summary: str
    overall_score: int = Field(ge=1, le=100)
    timeline: list[TimelineSegment]
    top_regions: list[RegionInsight]
    recommendations: list[Recommendation]
    strengths: list[str]
    weaknesses: list[str]


class AnalysisResultResponse(BaseModel):
    analysis_id: str
    status: AnalysisStatus
    created_at: datetime
    completed_at: datetime | None = None
    video_metadata: VideoMetadataSchema
    report: AnalysisReport
    brain_images: list[str]


# --- API Envelope ---


class ApiResponse(BaseModel):
    success: bool = True
    data: dict | AnalysisCreateResponse | AnalysisStatusResponse | AnalysisResultResponse | None = None
    error: str | None = None

File: frontend/src/types/analysis.ts
export type AnalysisStatus = 'queued' | 'processing' | 'completed' | 'failed'

export type ProcessingStage =
  | 'validating'
  | 'extracting_events'
  | 'tribe_predict'
  | 'summarizing_rois'
  | 'generating_plots'
  | 'analyzing_with_claude'
  | 'saving_results'

export interface AnalysisCreateResponse {
  readonly analysis_id: string
  readonly status: AnalysisStatus
  readonly created_at: string
}

export interface AnalysisStatusResponse {
  readonly analysis_id: string
  readonly status: AnalysisStatus
  readonly stage?: ProcessingStage
  readonly stage_label?: string
  readonly progress_percent: number
}

export interface VideoMetadata {
  readonly filename: string
  readonly duration_seconds: number
  readonly resolution: string
  readonly file_size_bytes: number
}

export interface TimelineSegment {
  readonly start_seconds: number
  readonly end_seconds: number
  readonly label: string
  readonly dominant_regions: readonly string[]
  readonly cognitive_state: string
  readonly engagement_level: 'high' | 'medium' | 'low'
  readonly insight: string
}

export interface RegionInsight {
  readonly roi_name: string
  readonly full_name: string
  readonly cognitive_function: string
  readonly activation_rank: number
  readonly marketing_implication: string
}

export interface Recommendation {
  readonly category: string
  readonly title: string
  readonly description: string
  readonly priority: 'high' | 'medium' | 'low'
}

export interface AnalysisReport {
  readonly executive_summary: string
  readonly overall_score: number
  readonly timeline: readonly TimelineSegment[]
  readonly top_regions: readonly RegionInsight[]
  readonly recommendations: readonly Recommendation[]
  readonly strengths: readonly string[]
  readonly weaknesses: readonly string[]
}

export interface AnalysisResult {
  readonly analysis_id: string
  readonly status: AnalysisStatus
  readonly created_at: string
  readonly completed_at?: string
  readonly video_metadata: VideoMetadata
  readonly report: AnalysisReport
  readonly brain_images: readonly string[]
}

export interface ApiResponse<T> {
  readonly success: boolean
  readonly data?: T
  readonly error?: string
}
